### PR TITLE
add sq8 classical quantizer and simd operators

### PR DESCRIPTION
**Pull request by LHT129**

### CODE REVIEW

File: src/simd/avx.cpp [hunk 1]
```diff
@@ -207,6 +210,182 @@ FP32ComputeL2Sqr(const float* query, const float* codes, uint64_t dim) {
...
+        __m256 diff_values = _mm256_loadu_ps(diff + i);
+        __m256 lowerBound_values = _mm256_loadu_ps(lowerBound + i);
+        // Perform calculations
+        __m256 scaled_codes1 = _mm256_fmadd_ps(codes1_floats, diff_values, lowerBound_values);
+        __m256 scaled_codes2 = _mm256_fmadd_ps(codes2_floats, diff_values, lowerBound_values);
+        __m256 val = _mm256_sub_ps(scaled_codes1, scaled_codes2);
+        val = _mm256_mul_ps(val, val);
+        sum = _mm256_add_ps(sum, val);
+    }
+    // Horizontal addition
```
Comment (ShawnShawnYou) [inline src/simd/avx.cpp:371]: redundant line

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -1,17 +1,26 @@
 
 add_subdirectory (simd)
+set (SIMD_QUANT_SRCS quantization/sq4_simd.cpp quantization/sq8_simd.cpp quantization/fp32_simd.cpp quantization/sq4_uniform_simd.cpp)
+set_source_files_properties (${SIMD_QUANT_SRCS} PROPERTIES COMPILE_FLAGS "-msse -msse2 -msse3 -mssse3 -msse4 -msse4a -msse4.1 -msse4.2")
+set_source_files_properties (${SIMD_QUANT_SRCS} PROPERTIES COMPILE_FLAGS "-mavx")
+set_source_files_properties (${SIMD_QUANT_SRCS} PROPERTIES COMPILE_FLAGS "-mavx2 -mfma")
+set_source_files_properties (${SIMD_QUANT_SRCS} PROPERTIES COMPILE_FLAGS "-mavx512f -mavx512pf -mavx512er -mavx512cd -mavx512vl -mavx512bw -mavx512dq -mavx512ifma -mavx512vbmi")
```
Comment (wxyucs) [inline src/CMakeLists.txt]: @LHT129 what's the behavior when compiler not support `sse` or some other flag?

File: src/quantization/fp32_quantizer.h [hunk 3]
```diff
@@ -118,8 +118,8 @@ template <MetricType Metric>
 void
 FP32Quantizer<Metric>::ProcessQueryImpl(const DataType* query,
                                         Computer<FP32Quantizer<Metric>>& computer) const {
-    computer.buf_ = new uint8_t[this->codeSize_];
-    memcpy(computer.buf_, query, this->codeSize_);
+    computer.buf_ = new uint8_t[this->code_size_];
```
Comment (inabao) [inline src/quantization/fp32_quantizer.h:121]: why not allocate the memory by our allocator

File: src/quantization/sq8_quantizer.h [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    } else {
+        return 0.;
+    }
+}
+
+template <MetricType Metric>
+void
+SQ8Quantizer<Metric>::ProcessQueryImpl(const DataType* query,
+                                       Computer<SQ8Quantizer>& computer) const {
+    computer.buf_ = new uint8_t[this->dim_ * sizeof(float)];
```
Comment (inabao) [inline src/quantization/sq8_quantizer.h:160]: Can use alloctor here?

File: src/simd/avx.cpp (hunk 1 above)
Comment (LHT129) [reply src/simd/avx.cpp:371]: done

File: src/CMakeLists.txt (hunk 2 above)
Comment (LHT129) [reply src/CMakeLists.txt]: delete

**Review by wxyucs — APPROVED** [review]
lgtm

**Pull request merged**